geomesa-convert/geomesa-convert-common/src/main/scala/org/locationtech/geomesa/convert2/AbstractConverter.scala

```diff
@@ -0,0 +1,291 @@
+import java.nio.charset.Charset
+
+import com.typesafe.config.Config
+import com.typesafe.scalalogging.LazyLogging
+import org.geotools.factory.Hints
+import org.locationtech.geomesa.convert.ErrorMode.ErrorMode
+import org.locationtech.geomesa.convert.ParseMode.ParseMode
+import org.locationtech.geomesa.convert._
+import org.locationtech.geomesa.convert2.transforms.Expression
+import org.locationtech.geomesa.features.ScalaSimpleFeature
+import org.locationtech.geomesa.utils.collection.CloseableIterator
+import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
+
+import scala.collection.mutable.{ArrayBuffer, ListBuffer}
+import scala.util.control.NonFatal
+
+/**
+  * Abstract converter implementation. Typically paired with an AbstractConverterFactory. If so, needs to have
+  * a default constructor consisting of (targetSft, config, fields, options), so that the AbstractConverterFactory
+  * can instantiate it via reflection.
+  *
+  * Subclasses need to implement `read` to parse the underlying input stream into raw values that will be
+  * transformed to simple features.
+  *
+  * @param targetSft simple feature type
+  * @param config converter config
+  * @param fields converter fields
+  * @param options converter options
+  * @tparam C config binding
+  * @tparam F field binding
+  * @tparam O options binding
+  */
+abstract class AbstractConverter[C <: ConverterConfig, F <: Field, O <: ConverterOptions]
+  (override val targetSft: SimpleFeatureType, val config: C, fields: Seq[F], options: O)
+    extends SimpleFeatureConverter with LazyLogging {
+
+  private val requiredFields: Array[Field] =
+    AbstractConverter.requiredFields(targetSft, fields, config.userData.values.toSeq ++ config.idField.toSeq)
+
+  private val requiredFieldsCount: Int = requiredFields.length
+
+  private val requiredFieldsIndices: Array[Int] = requiredFields.map(f => targetSft.indexOf(f.name))
+
+  /**
+    * Read values for simple features out of the input stream. This should be lazily evaluated,
+    * so that any exceptions occur in the call to `hasNext`. If there is any sense of 'lines',
+    * they should be indicated with `ec.counter.incLineCount`
+    *
+    * @param is input
+    * @param ec evaluation context
+    * @return raw extracted values, one iterator entry per simple feature
+    */
+  protected def read(is: InputStream, ec: EvaluationContext): CloseableIterator[Array[Any]]
+
+  override def createEvaluationContext(globalParams: Map[String, Any],
+                                       caches: Map[String, EnrichmentCache],
+                                       counter: Counter): EvaluationContext = {
+    val globalKeys = globalParams.keys.toSeq
+    val names = requiredFields.map(_.name) ++ globalKeys
+    val values = Array.ofDim[Any](names.length)
+    // note, globalKeys are maintained even through EvaluationContext.clear()
+    globalKeys.zipWithIndex.foreach { case (k, i) => values(requiredFields.length + i) = globalParams(k) }
```

Comment:
Not that it matters much, but what about:

```scala
val (globalKeys, values) = globalParams.toArray.foldLeft((Array[Any](), Array[Any]()))((acc,t) => (acc._1 :+ t._1, acc._2 :+ t._2))
```

This yields two variables of type `Array[Any]` without the trouble of `zipWithIndex`.

Reply:
i replaced `zipWithIndex.foreach` with `org.locationtech.geomesa.utils.conversions.ScalaImplicits.RichTraversableOnce.foreachIndex`, which uses a while loop for optimal efficiency